spring-xd-messagebus-rabbit/src/main/java/org/springframework/xd/dirt/integration/rabbit/RabbitMessageBus.java

```diff
@@ -673,6 +677,34 @@ public void destroy() {
 		stopBindings();
 	}
 
+	@Override
+	public void storeForManualAck(MessageHeaders headers) {
+		if (headers.containsKey(AmqpHeaders.CHANNEL)) {
+			Channel channel = (com.rabbitmq.client.Channel) headers.get(AmqpHeaders.CHANNEL);
+			Long deliveryTag = (Long) headers.get(AmqpHeaders.DELIVERY_TAG);
+			if (channelsForManualAck.containsKey(channel) && (channelsForManualAck.get(channel) < deliveryTag)) {
+				channelsForManualAck.put(channel, deliveryTag);
+			}
+			else {
+				channelsForManualAck.put(channel, deliveryTag);
+			}
+		}
+	}
+
+	@Override
+	public void doManualAck() {
+		for (Map.Entry<Object, Long> entry : channelsForManualAck.entrySet()) {
```

Comment:
AFAICT this will only work if concurrency = 1 on the bus (and then there is only one channel).

Consider concurrency = 2.

Correct me if I'm wrong; let's say the batch size is 10 each thread delivers 5 messages and this triggers a push. The spark thread starts to call this method but, in the meantime, message 6 on thread 1 arrives, updating its deliveryTag to 6. Won't the spark push then ack 6 for thread 1 and 5 for thread 2 (which is wrong)?

It seems to me you need to keep a bucket of acks somehow and then figure out at ack time which ones were actually part of the block. Do you have access to the streamBlockId that receives a message?

Am I missing something?